cpp/src/strings/split/split.cu

```diff
@@ -385,6 +386,348 @@ struct whitespace_rsplit_tokenizer_fn
+            d_str.byte_offset(end_pos) -
+              d_str.byte_offset(delimiter_pos + d_delimiter.length());
+        token_count++;
+        token_size_sum += token_char_bytes;
+        if (forward) {
+          start_pos = delimiter_pos + d_delimiter.length();
+        }
+        else {
+          end_pos = delimiter_pos;
+        }
+      }
+      else {
+        break;
+      }
+    }
+    token_count++;
+    token_size_sum +=
+      forward ? d_str.byte_offset(end_pos) - d_str.byte_offset(start_pos) :
+                d_str.byte_offset(end_pos);
+
+    auto memory_size =
+      round_up_pow2(token_size_sum, split_align) +
+      round_up_pow2(
+        (token_count + 1) * static_cast<size_type>(sizeof(size_type)),
+        split_align);
+
+    return thrust::make_tuple<size_type, size_type, size_type>(
+      token_count, token_size_sum, memory_size);
+  }
+};
+
+/**
+ * @brief Copy the tokens from the `idx'th` string element of `d_strings` to
+ * the contiguous memory buffer.
+ */
+template <bool forward>
+struct token_copier_fn {
+  column_device_view const d_strings;  // strings to split
+  string_view const d_delimiter;  // delimiter for split
+  bool const has_validity = false;
+
+  __device__ void
+  operator()(
+      thrust::tuple<size_type, size_type, void*, size_type>
+        idx_token_count_memory_ptr_token_size_sum) const {
+    auto idx = idx_token_count_memory_ptr_token_size_sum.get<0>();
+    auto token_count = idx_token_count_memory_ptr_token_size_sum.get<1>();
+
+    if (token_count == 0) {
+      return;
+    }
+
+    auto memory_ptr =
+      static_cast<char*>(idx_token_count_memory_ptr_token_size_sum.get<2>());
+    auto token_size_sum = idx_token_count_memory_ptr_token_size_sum.get<3>();
+    auto char_buf_size = round_up_pow2(token_size_sum, split_align);
+
+    auto char_buf_ptr = memory_ptr;
+    memory_ptr += char_buf_size;
+    auto offset_buf_ptr = reinterpret_cast<size_type*>(memory_ptr);
+
+    auto d_str = d_strings.element<string_view>(idx);
+    size_type token_idx = 0;
+    size_type char_bytes_copied = 0;
+    size_type start_pos = 0;  // updates only if forward is true
+    auto end_pos = d_str.length();  // updates only if forward is false
+    while (token_idx < token_count - 1) {
+      auto delimiter_pos =
+        forward ? d_str.find(d_delimiter, start_pos) :
+                  d_str.rfind(d_delimiter, start_pos, end_pos);
+      if (delimiter_pos != -1) {
+        auto src_byte_offset =
+          forward ? d_str.byte_offset(start_pos) :
+                    d_str.byte_offset(delimiter_pos + d_delimiter.length());
+        auto token_char_bytes =
+          forward ?
+            d_str.byte_offset(delimiter_pos) - src_byte_offset :
+            d_str.byte_offset(end_pos) - src_byte_offset;
+        if (forward) {
+          for (size_type i = 0; i < token_char_bytes; ++i) {
+            char_buf_ptr[char_bytes_copied + i] =
+              d_str.data()[src_byte_offset + i];
+          }
```

Comment:
Can all these `for` loops be `std::copy`s?

Reply:
If I am not mistaken, calling std::copy within a device function did not compile with the following error messages.

```
[ 12%] Building CUDA object CMakeFiles/cudf.dir/src/strings/split/split.cu.o
/home/seunghwak/RAPIDS/cudf/cpp/src/strings/split/split.cu(532): error: calling a __host__ function("std::copy<const char *, char *> ") from a __device__ function("cudf::strings::detail::_NV_ANON_NAMESPACE::token_copier_fn<(bool)1> ::operator () const") is not allowed

/home/seunghwak/RAPIDS/cudf/cpp/src/strings/split/split.cu(532): error: identifier "std::copy<const char *, char *> " is undefined in device code

/home/seunghwak/RAPIDS/cudf/cpp/src/strings/split/split.cu(532): error: calling a __host__ function("std::copy<const char *, char *> ") from a __device__ function("cudf::strings::detail::_NV_ANON_NAMESPACE::token_copier_fn<(bool)0> ::operator () const") is not allowed

/home/seunghwak/RAPIDS/cudf/cpp/src/strings/split/split.cu(532): error: identifier "std::copy<const char *, char *> " is undefined in device code
```

I also thought about using memcpy instead, but it seems like there is a performance issue related to this (https://stackoverflow.com/questions/10456728/is-there-an-equivalent-to-memcpy-that-works-inside-a-cuda-kernel).

Reply:
You can call `thrust::copy` with `thrust::sequential` policy to avoid dynamic parallelism.

Reply:
Here's an example:
```
  thrust::transform(thrust::make_counting_iterator<size_t>(0), thrust::make_counting_iterator<size_t>(n),
            d_row_offsets.begin(),
            [I, n_entries] __device__ (size_t row_indx){
                      ///return thrust::count(thrust::device, I, I+n_entries, row_indx);//<-don't do this: dynamic para: slow
                      return thrust::count(thrust::seq, I, I+n_entries, row_indx);//do this instead
            });
```

Reply:
Cool, this worked!